src/lib.rs

```diff
@@ -1,14 +1,207 @@
-pub fn add(left: usize, right: usize) -> usize {
-    left + right
+use std::{cmp::Ordering, hash};
+
+pub use proc_macros::grouped_ordering;
+
+pub trait GroupedOrdering {
+    type Group: hash::Hash + Eq;
+
+    fn compare(&self, a: &Self::Group, b: &Self::Group) -> Ordering;
+}
+
+pub trait GroupedOrderable<TGroupedOrdering: GroupedOrdering> {
+    fn map_to_grouped_ordering(&self) -> TGroupedOrdering::Group;
+}
+
+pub trait VecExt {
+    type Item;
+
+    fn sort_by_grouped_ordering<TGroupedOrdering>(&mut self, grouped_ordering: &TGroupedOrdering)
```

Comment:
This seemed like the most obvious "use case" to try and expose a "nice" API for but I'm guessing eg there would be other collection-sorting API's that should be mimicked/implemented?